epil example model does not compile on Mac OS X 10.7.5

Starting with a clean install:

```
make clean-all
make bin/stanc
```

Then running some of the tests:

```
make src/test/CmdStan
```

The above command eventually fails:

```
--- Copying model ---
cp src/models/bugs_examples/vol1/epil/epil.stan models/bugs_examples/vol1/epil/epil.stan 

--- Copying data ---
cp src/models/bugs_examples/vol1/epil/epil.data.R models/bugs_examples/vol1/epil/epil.data.R

--- Translating Stan model to C++ code ---
bin/stanc models/bugs_examples/vol1/epil/epil.stan --o=models/bugs_examples/vol1/epil/epil.cpp
Model name=epil_model
Input file=models/bugs_examples/vol1/epil/epil.stan
Output file=models/bugs_examples/vol1/epil/epil.cpp
g++ -I src -isystem lib/eigen_3.2.0 -isystem lib/boost_1.54.0 -Wall -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -pipe -Wno-unused-function -ftemplate-depth-256   -c -O3 -o models/bugs_examples/vol1/epil/epil.o models/bugs_examples/vol1/epil/epil.cpp
models/bugs_examples/vol1/epil/epil.cpp: In member function ‘T epil_model_namespace::epil_model::log_prob(Eigen::Matrix<T_cut, -0x00000000000000001, 1, 0, -0x00000000000000001, 1>&, std::ostream*) const’:
models/bugs_examples/vol1/epil/epil.cpp:493: error: ‘epil_model_namespace::epil_model::T’ cannot appear in a constant-expression
models/bugs_examples/vol1/epil/epil.cpp:493: error: template argument 1 is invalid
models/bugs_examples/vol1/epil/epil.cpp:493: error: template argument 2 is invalid
models/bugs_examples/vol1/epil/epil.cpp:493: error: invalid type in declaration before ‘;’ token
models/bugs_examples/vol1/epil/epil.cpp:494: error: request for member ‘reserve’ in ‘vec_params_r’, which is of non-class type ‘int’
models/bugs_examples/vol1/epil/epil.cpp:496: error: request for member ‘push_back’ in ‘vec_params_r’, which is of non-class type ‘int’
models/bugs_examples/vol1/epil/epil.cpp:498: error: ‘epil_model_namespace::epil_model::T’ cannot appear in a constant-expression
src/stan/math/constants.hpp: At global scope:
src/stan/math/constants.hpp:63: warning: ‘stan::math::NEGATIVE_EPSILON’ defined but not used
make: *** [models/bugs_examples/vol1/epil/epil.o] Error 1
```

Looking at the code mentioned in the error message, it appears to be a compiler bug of some sort. Here is the code where the problem occurs, with line 493 being the declaration of vec_params_r:

```
    template <bool propto, bool jacobian, typename T>
    T log_prob(Eigen::Matrix<T,Eigen::Dynamic,1>& params_r,
               std::ostream* pstream = 0) const {
      std::vector<T> vec_params_r;
      vec_params_r.reserve(params_r.size());
      for (int i = 0; i < params_r.size(); ++i)
        vec_params_r.push_back(params_r(i));
      std::vector<int> vec_params_i;
      return log_prob<propto,jacobian,T>(vec_params_r, vec_params_i, pstream);
    }
```

Apparently the compiler is confusing the typename T with an int member variable of the same name.

Some more information:

```
mac> which g++
/usr/bin/g++
mac> g++ --version
i686-apple-darwin11-llvm-g++-4.2 (GCC) 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.11.00)
Copyright (C) 2007 Free Software Foundation, Inc.
```

--- a/src/stan/gm/generator.hpp
+++ b/src/stan/gm/generator.hpp
@@ -1651,15 +1651,15 @@ namespace stan {
       o << INDENT2 << "return lp_accum__.sum();" << EOL2;
       o << INDENT << "} // log_prob()" << EOL2;
 
-      o << INDENT << "template <bool propto, bool jacobian, typename T>" << EOL;
-      o << INDENT << "T log_prob(Eigen::Matrix<T,Eigen::Dynamic,1>& params_r," << EOL;
+      o << INDENT << "template <bool propto, bool jacobian, typename T_>" << EOL;
+      o << INDENT << "T_ log_prob(Eigen::Matrix<T_,Eigen::Dynamic,1>& params_r," << EOL;
       o << INDENT << "           std::ostream* pstream = 0) const {" << EOL;
-      o << INDENT << "  std::vector<T> vec_params_r;" << EOL;
+      o << INDENT << "  std::vector<T_> vec_params_r;" << EOL;
       o << INDENT << "  vec_params_r.reserve(params_r.size());" << EOL;
       o << INDENT << "  for (int i = 0; i < params_r.size(); ++i)" << EOL;
       o << INDENT << "    vec_params_r.push_back(params_r(i));" << EOL;
       o << INDENT << "  std::vector<int> vec_params_i;" << EOL;
-      o << INDENT << "  return log_prob<propto,jacobian,T>(vec_params_r, vec_params_i, pstream);" << EOL;
+      o << INDENT << "  return log_prob<propto,jacobian,T_>(vec_params_r, vec_params_i, pstream);" << EOL;
       o << INDENT << "}" << EOL2;
     }
 

